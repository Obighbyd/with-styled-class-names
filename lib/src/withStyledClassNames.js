var _templateObject = _taggedTemplateLiteral(['\n      ', ';\n    '], ['\n      ', ';\n    ']),
    _templateObject2 = _taggedTemplateLiteral([''], ['']);

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

import React from 'react';
import styled from 'styled-components';

var withCustomClassNameProp = function withCustomClassNameProp(Component, propClass, keyToNest) {
  return function (_ref) {
    var className = _ref.className,
        rest = _objectWithoutProperties(_ref, ['className']);

    var classNamesAlreadySet = keyToNest ? rest[keyToNest] ? rest[keyToNest] : {} : undefined;
    var objToMerge = keyToNest ? _defineProperty({}, propClass, className) : undefined;
    var merged = Object.assign({}, classNamesAlreadySet, objToMerge);
    var props = keyToNest ? Object.assign({}, rest, _defineProperty({}, keyToNest, merged)) : Object.assign({}, rest, _defineProperty({}, propClass, className));
    return React.createElement(Component, props);
  };
};

var injectControlledClassNames = function injectControlledClassNames(stylesMap, Component, keyToNest) {
  return Object.entries(stylesMap).reduce(function (ResultComponent, entry) {
    var customClassNameProp = entry[0];
    var tagOrStyledComponent = entry[1];
    if (tagOrStyledComponent && tagOrStyledComponent.name === 'StyledComponent') {
      return tagOrStyledComponent.withComponent(withCustomClassNameProp(ResultComponent, customClassNameProp, keyToNest));
    }
    if (tagOrStyledComponent && !(tagOrStyledComponent instanceof Array) && typeof tagOrStyledComponent !== 'string') {
      return injectControlledClassNames(tagOrStyledComponent, ResultComponent, customClassNameProp);
    }
    return styled(withCustomClassNameProp(ResultComponent, customClassNameProp, keyToNest))(_templateObject, tagOrStyledComponent);
  }, Component);
};

function withStyledClassNames(stylesMap, Component) {
  var isComponentPassed = !!Component;
  return isComponentPassed ? injectControlledClassNames(stylesMap, styled(Component)(_templateObject2)) : function (Component) {
    return injectControlledClassNames(stylesMap, styled(Component)(_templateObject2));
  };
}

export default withStyledClassNames;