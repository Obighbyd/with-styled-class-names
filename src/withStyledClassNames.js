import React from 'react'
import styled from 'styled-components'

const withCustomClassNameProp = (Component, propClass, keyToNest) => ({
  className,
  ...rest
}) => {
  const classNamesAlreadySet = keyToNest
    ? rest[keyToNest] ? rest[keyToNest] : {}
    : undefined
  const objToMerge = keyToNest ? { [propClass]: className } : undefined
  const merged = { ...classNamesAlreadySet, ...objToMerge }
  const props = keyToNest
    ? { ...rest, [keyToNest]: merged }
    : { ...rest, [propClass]: className }
  return <Component {...props} />
}

const injectControlledClassNames = (stylesMap, Component, keyToNest) =>
  Object.entries(stylesMap).reduce((ResultComponent, entry) => {
    const customClassNameProp = entry[0]
    const tagOrStyledComponent = entry[1]
    if (
      tagOrStyledComponent &&
      tagOrStyledComponent.name === 'StyledComponent'
    ) {
      return tagOrStyledComponent.withComponent(
        withCustomClassNameProp(
          ResultComponent,
          customClassNameProp,
          keyToNest,
        ),
      )
    }
    if (
      tagOrStyledComponent &&
      !(tagOrStyledComponent instanceof Array) &&
      typeof tagOrStyledComponent !== 'string'
    ) {
      return injectControlledClassNames(
        tagOrStyledComponent,
        ResultComponent,
        customClassNameProp,
      )
    }
    return styled(
      withCustomClassNameProp(ResultComponent, customClassNameProp, keyToNest),
    )`
      ${tagOrStyledComponent};
    `
  }, Component)

function withStyledClassNames(stylesMap, Component) {
  const isComponentPassed = !!Component
  return isComponentPassed
    ? injectControlledClassNames(stylesMap, styled(Component)``)
    : Component => {
        return injectControlledClassNames(stylesMap, styled(Component)``)
      }
}

export default withStyledClassNames
