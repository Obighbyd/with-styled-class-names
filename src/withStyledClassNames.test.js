import React from 'react'
import styled, { css } from 'styled-components'
import { shallow, mount } from 'enzyme'
import withStyledClassNames from './withStyledClassNames'

const Component = ({
  className,
  nestedClassName,
  anotherNestedClassName,
  objectClassNamesList = { first: {}, second: {} },
  ...props
}) => {
  const { first, second } = objectClassNamesList
  return (
    <div id="root" className={className} {...props}>
      {'Root'}
      <div id="nestedClassName" className={nestedClassName}>
        {'Nested One Level'}
        <div id="anotherNestedClassName" className={anotherNestedClassName}>
          Another nested
        </div>
        <div id="first" className={first}>
          Another first
        </div>
        <div id="second" className={second}>
          Another second
        </div>
      </div>
    </div>
  )
}

const DeriveStyles = styled.div`
  background: aliceblue;
`

describe('withStyledClassNames', () => {
  test('Style container', () => {
    const WithClassNames = withStyledClassNames({}, Component)
      .extend`background: red;`
    const rendered = shallow(<WithClassNames />)
    expect(rendered).toHaveStyleRule('background', 'red')
  })

  test('Style container and curried', () => {
    const WithClassNames = withStyledClassNames({})(Component)
      .extend`background: red;`
    const rendered = shallow(<WithClassNames />)
    expect(rendered).toHaveStyleRule('background', 'red')
  })

  test('Style container by deriving styles of other styled-component', () => {
    const WithClassNames = DeriveStyles.withComponent(
      withStyledClassNames({}, Component),
    )
    const rendered = mount(<WithClassNames />)
    expect(rendered).toHaveStyleRule('background', 'aliceblue')
  })

  test('To not override className if already provided', () => {
    const WithClassNames = withStyledClassNames(
      {
        className: `background: red;`,
      },
      Component,
    )
    const rendered = mount(<WithClassNames />).find('#root')
    expect(rendered).toHaveProp('className')
    expect(rendered).toHaveStyleRule('background', 'red')
  })

  test('To provide nested props', () => {
    const WithClassNames = withStyledClassNames(
      {
        nestedClassName: `background: red;`,
      },
      Component,
    )
    const rendered = mount(<WithClassNames />)
    const nested = rendered.find('#nestedClassName')
    expect(nested).toHaveProp('className')
    expect(nested).toHaveStyleRule('background', 'red')
  })

  test('To provide one level nested props', () => {
    const WithClassNames = withStyledClassNames(
      {
        objectClassNamesList: {
          first: `background: yellow`,
          second: `background: purple`,
        },
      },
      Component,
    )
    const rendered = mount(<WithClassNames />)
    const first = rendered.find('#first')
    const second = rendered.find('#second')
    expect(first).toMatchSnapshot()
    expect(second).toMatchSnapshot()
    expect(first).toHaveProp('className')
    expect(first).toHaveStyleRule('background', 'yellow')
    expect(second).toHaveProp('className')
    expect(second).toHaveStyleRule('background', 'purple')
  })

  test('All levels could be styled by another styled-component', () => {
    const WithClassNames = withStyledClassNames(
      {
        nestedClassName: styled.div`
          background: red;
        `,
        objectClassNamesList: {
          first: styled.div`
            background: yellow;
          `,
          second: styled.div`
            background: purple;
          `,
        },
      },
      Component,
    )
    const rendered = mount(<WithClassNames />)
    const nestedClassName = rendered.find('#nestedClassName')
    const first = rendered.find('#first')
    const second = rendered.find('#second')
    expect(nestedClassName).toMatchSnapshot()
    expect(first).toMatchSnapshot()
    expect(second).toMatchSnapshot()
    expect(nestedClassName).toHaveProp('className')
    expect(nestedClassName).toHaveStyleRule('background', 'red')
    expect(first).toHaveProp('className')
    expect(first).toHaveStyleRule('background', 'yellow')
    expect(second).toHaveProp('className')
    expect(second).toHaveStyleRule('background', 'purple')
  })

  test('All levels props dependent styling working', () => {
    const WithClassNames = withStyledClassNames(
      {
        nestedClassName: css`
          ${p => p.bool && 'background: red;'};
        `,
        objectClassNamesList: {
          first: css`
            ${p => p.bool && 'background: yellow;'};
          `,
          second: css`
            ${p => p.bool && 'background: purple;'};
          `,
        },
      },
      Component,
    )

    const rendered = mount(<WithClassNames bool />)
    const nestedClassName = rendered.find('#nestedClassName')
    const first = rendered.find('#first')
    const second = rendered.find('#second')
    expect(nestedClassName).toMatchSnapshot()
    expect(first).toMatchSnapshot()
    expect(second).toMatchSnapshot()
    expect(nestedClassName).toHaveProp('className')
    expect(nestedClassName).toHaveStyleRule('background', 'red')
    expect(first).toHaveProp('className')
    expect(first).toHaveStyleRule('background', 'yellow')
    expect(second).toHaveProp('className')
    expect(second).toHaveStyleRule('background', 'purple')
  })
})
