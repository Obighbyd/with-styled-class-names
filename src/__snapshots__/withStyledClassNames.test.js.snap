// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withStyledClassNames All levels could be styled by another styled-component 1`] = `
.c0 {
  background: red;
}

.c1 {
  background: yellow;
}

.c2 {
  background: purple;
}

<div
  className="c0"
  id="nestedClassName"
>
  Nested One Level
  <div
    className={undefined}
    id="anotherNestedClassName"
  >
    Another nested
  </div>
  <div
    className="c1"
    id="first"
  >
    Another first
  </div>
  <div
    className="c2"
    id="second"
  >
    Another second
  </div>
</div>
`;

exports[`withStyledClassNames All levels could be styled by another styled-component 2`] = `
.c0 {
  background: yellow;
}

<div
  className="c0"
  id="first"
>
  Another first
</div>
`;

exports[`withStyledClassNames All levels could be styled by another styled-component 3`] = `
.c0 {
  background: purple;
}

<div
  className="c0"
  id="second"
>
  Another second
</div>
`;

exports[`withStyledClassNames All levels props dependent styling working 1`] = `
.c0 {
  background: red;
}

.c1 {
  background: yellow;
}

.c2 {
  background: purple;
}

<div
  className="c0"
  id="nestedClassName"
>
  Nested One Level
  <div
    className={undefined}
    id="anotherNestedClassName"
  >
    Another nested
  </div>
  <div
    className="c1"
    id="first"
  >
    Another first
  </div>
  <div
    className="c2"
    id="second"
  >
    Another second
  </div>
</div>
`;

exports[`withStyledClassNames All levels props dependent styling working 2`] = `
.c0 {
  background: yellow;
}

<div
  className="c0"
  id="first"
>
  Another first
</div>
`;

exports[`withStyledClassNames All levels props dependent styling working 3`] = `
.c0 {
  background: purple;
}

<div
  className="c0"
  id="second"
>
  Another second
</div>
`;

exports[`withStyledClassNames To provide one level nested props 1`] = `
.c0 {
  background: yellow;
}

<div
  className="c0"
  id="first"
>
  Another first
</div>
`;

exports[`withStyledClassNames To provide one level nested props 2`] = `
.c0 {
  background: purple;
}

<div
  className="c0"
  id="second"
>
  Another second
</div>
`;
